### ISSUE TITLE
Loss of type information after variable assignment

### ISSUE THREAD

**Issue by skogsbaer** [etylizer/etylizer#53]
Consider the following program:

~~~erlang
-spec bar(boolean()) -> boolean().
bar(X) -> X.

-spec foo(boolean()) -> boolean().
foo(B) ->           
    B2 = B,          
    case bar(B2) of
        _ -> B2
    end.
~~~

It should typecheck but it fails!

The problem is function `foo`. Internally, we treat the assignment `B2 = B` as a case, similar as in the following version of foo, where I also added type information for the variables `B` and `B2`:

~~~erlang
-spec foo(boolean()) -> boolean().
foo(B) ->                 % B has type boolean()
    case B of
        B2 ->                % B2 has type boolean() | mu for some fresh tyvar mu
          case bar(B2) of
              _ -> B2
          end
    end.
~~~

What happens here is the following:

- scrutiny `B` (first case) generates the constraint `B <= alpha` for some fresh tyvar alpha
- the branch `case bar(B2) of ...` is then typed in the environment `{B: alpha /\ boolean(), B2: alpha}`
- Solving the constraint `B <= alpha` results in substitution `{alpha -> boolean() | mu` for fresh mu
- Hence, `B2` has type `boolean() | mu` in the branch `case bar(B2) of ...`
- Typing the invocation `bar(B2)` requires solving the constraint `boolean() | mu <= boolean()`, which has no solution because `mu` is free in the environment and so fixed for tally.

This problem seems to be very related to bug #36. We would want the constraint
`B ~ alpha` for the scrutiny so that the substitution becomes `{alpha -> boolean()}`.

A somewhat related test case:

~~~erlang
-spec foo(boolean()) -> integer().
foo(B) ->
    B2 = not B,
    I = if B andalso B2 -> 1; true -> 2 end,
    I.
~~~

**Comment by skogsbaer** [comment]
The following program type checks:

~~~erlang
-spec foo(boolean()) -> boolean().
foo(B) ->
    B2 = B,
    B2.
~~~

Typing it also requires solving the constraint `boolean() | mu <= boolean()`, but this time without `mu` in the set of variables fixed for tally.

**Comment by skogsbaer** [comment]
I think the solution is to use an empty set of fixed variables when invoking tally in rule RC-CASE.

**Comment by skogsbaer** [comment]
fixed in sw/error-msgs (PR #38)
